viral/rastermap_utils.py

```diff
@@ -0,0 +1,415 @@
+import numpy as np
+import sys
+import random
+from pathlib import Path
+from scipy.interpolate import interp1d
+from typing import List
+
+HERE = Path(__file__).parent
+sys.path.append(str(HERE.parent))
+sys.path.append(str(HERE.parent.parent))
+
+from viral.utils import (
+    TrialInfo,
+    trial_is_imaged,
+    degrees_to_cm,
+    get_wheel_circumference_from_rig,
+)
+from viral.models import Cached2pSession
+from viral.constants import TIFF_UMBRELLA
+from viral.two_photon import (
+    compute_dff,
+    subtract_neuropil,
+    normalize,
+)
+
+
+def get_spks_pos(s2p_path: str) -> tuple[np.ndarray]:
+    iscell = np.load(s2p_path / "iscell.npy")[:, 0].astype(bool)
+    spks = np.load(s2p_path / "spks.npy")[iscell, :]
+    stat = np.load(s2p_path / "stat.npy", allow_pickle=True)[iscell]
+    pos = np.array([[int(coord) for coord in cell["med"]] for cell in stat])
+    xpos = pos[:, 0]
+    ypos = pos[:, 1]
+    return spks, xpos, ypos
+
+
+def get_dff(s2p_path: str) -> np.ndarray:
+    iscell = np.load(s2p_path / "iscell.npy")[:, 0].astype(bool)
+    f_raw = np.load(s2p_path / "F.npy")[iscell, :]
+    f_neu = np.load(s2p_path / "Fneu.npy")[iscell, :]
+    return compute_dff(subtract_neuropil(f_raw, f_neu))
+
+
+def align_trial_frames(trials: list[TrialInfo], ITI: bool = False) -> np.ndarray:
+    """Extract trial start and end frames, reward condition"""
+    trial_frames = np.array(
+        [
+            [trial.trial_start_closest_frame, trial.trial_end_closest_frame]
+            for trial in trials
+        ]
+    )
+    if ITI == False:
+        ITI_start_frames = np.array([get_ITI_start_frame(trial) for trial in trials])
+        trial_frames[:, 1] = ITI_start_frames
+    assert (
+        len(trials) == trial_frames.shape[0]
+    ), "Number of trials does not match number of trial times"
+    sorted_trial_frames = trial_frames[np.argsort(trial_frames[:, 0])]
+    for i in range(len(sorted_trial_frames) - 1):
+        assert (
+            sorted_trial_frames[i, 1] < sorted_trial_frames[i + 1, 0]
+        ), "Overlapping frames for trials"
+    rewarded = np.array([trial.texture_rewarded for trial in trials])
+    return np.column_stack((sorted_trial_frames, rewarded))
+
+
+def get_ITI_start_frame(trial: TrialInfo) -> float:
+    for state in trial.states_info:
+        if state.name == "ITI":
+            ITI_start_frame = state.closest_frame_start
+    return ITI_start_frame
+
+
+def get_signal_for_trials(
+    signal: np.ndarray, trial_frames: np.ndarray
+) -> list[np.ndarray]:
+    "Return spks or dff for trials"
+    signal_trials = list()
+    for trial in trial_frames:
+        trial_start, trial_end, _ = trial.astype(int)
+        signal_trials.append(signal[:, trial_start : trial_end + 1])
+    return signal_trials
+
+
+def get_frame_position(
+    trial: TrialInfo,
+    trial_frames: np.ndarray,
+    wheel_circumference: float,
+    ITI: bool = False,
+) -> np.ndarray:
+    if ITI == True:
```

Comment:
probably a dosctring here explaining what everything is. If you wana get really fancy you can use the [google suggested docstring style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)